lib/blockstore/file.js

```diff
@@ -0,0 +1,789 @@
+/*!
+ * blockstore/file.js - file blockstore for bcoin
+ * Copyright (c) 2019, Braydon Fuller (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const {isAbsolute, resolve, join} = require('path');
+const bdb = require('bdb');
+const assert = require('bsert');
+const fs = require('bfile');
+const bio = require('bufio');
+const Network = require('../protocol/network');
+const consensus = require('../protocol/consensus');
+const Headers = require('../primitives/headers');
+const AbstractBlockStore = require('./abstract');
+const {BlockRecord, FileRecord} = require('./records');
+const layout = require('./layout');
+const {types, prefixes} = require('./common');
+
+const WRITE = 0;
+const PRUNE = 1;
+
+/**
+ * File Block Store
+ *
+ * @alias module:blockstore:FileBlockStore
+ * @abstract
+ */
+
+class FileBlockStore extends AbstractBlockStore {
+  /**
+   * Create a blockstore that stores blocks in files.
+   * @constructor
+   */
+
+  constructor(options) {
+    super(options);
+
+    assert(isAbsolute(options.location), 'Location not absolute.');
+
+    this.location = options.location;
+    this.indexLocation = resolve(this.location, './index');
+
+    this.db = bdb.create({
+      location: this.indexLocation,
+      cacheSize: options.cacheSize,
+      compression: false
+    });
+
+    this.maxFileLength = options.maxFileLength || 128 * 1024 * 1024;
+
+    assert(Number.isSafeInteger(this.maxFileLength),
+      'Invalid max file length.');
+
+    this.network = Network.primary;
+
+    if (options.network != null)
+      this.network = Network.get(options.network);
+
+    this.writing = Object.create(null);
+  }
+
+  /**
+   * Compares the number of files in the directory
+   * with the recorded number of files.
+   * @param {Number} type - The type of block data
+   * @private
+   * @returns {Promise<Object>}
+   */
+
+  async check(type) {
+    const prefix = prefixes[type];
+    const regexp = new RegExp(`^${prefix}(\\d{5})\\.dat$`);
+    const all = await fs.readdir(this.location);
+    const dats = all.filter(f => regexp.test(f));
+    const filenos = dats.map(f => parseInt(f.match(regexp)[1]));
+
+    let missing = false;
+
+    for (const fileno of filenos) {
+      const rec = await this.db.get(layout.f.encode(type, fileno));
+      if (!rec) {
+        missing = true;
+        break;
+      }
+    }
+
+    return {missing, filenos};
+  }
+
+  /**
+   * Creates indexes from files for a block type. Reads the hash of
+   * the block data from the magic prefix, except for a block which
+   * the hash is read from the block header.
+   * @private
+   * @param {Number} type - The type of block data
+   * @returns {Promise}
+   */
+
+  async _index(type) {
```

Comment:
That is definitely something to look into, I would love to have block store as flexible as possible with different sources of corruption. I think for this PR keeping it how it is currently is okay and we can have discussion on new issue how to approach those things.

E.g. if 1 blk + blu is missing, should it only download those and skip the rest (How much rewrite would that require)
only blk or only blu is missing and etc. At least we could add check and fail early, instead of node getting stuck at certain height.

Reply:
Ok cool good plan.